File: app/src/payload.py
# Copyright (C) 2025 Logical Mechanism LLC
# SPDX-License-Identifier: GPL-3.0-only

# src/payload.py

import cbor2


def build_payload(
    locator: bytes,
    secret: bytes | None = None,
    digest: bytes | None = None,
    extra: dict[int, bytes] | None = None,
) -> bytes:
    """
    Build a canonical CBOR-encoded peace-payload map.

    The payload follows the peace-payload CDDL schema:
        { 0 => bstr, ? 1 => bstr, ? 2 => bstr, * int => bstr }

    Uses canonical CBOR encoding (RFC 8949 §4.2) for deterministic output
    across Python (cbor2) and TypeScript (cborg).

    Args:
        locator: Content address — IPFS CID, Arweave TX ID, URL, or inline data.
        secret: Optional access/decryption key for off-chain content.
        digest: Optional integrity hash of the underlying content.
        extra: Optional dict of extension fields (keys must be >= 3).

    Returns:
        Canonical CBOR-encoded bytes.

    Raises:
        ValueError: If extra contains reserved keys (0, 1, 2) or non-int keys.
    """
    m: dict[int, bytes] = {0: locator}
    if secret is not None:
        m[1] = secret
    if digest is not None:
        m[2] = digest
    if extra:
        for k, v in extra.items():
            if k in (0, 1, 2):
                raise ValueError(f"Extra key {k} conflicts with reserved keys (0, 1, 2)")
            m[k] = v
    return cbor2.dumps(m, canonical=True)


def parse_payload(data: bytes) -> dict[int, bytes]:
    """
    Parse a CBOR-encoded peace-payload map.

    Validates that the decoded value is a map with integer keys and
    byte string values, and that the required field 0 (locator) is present.

    Args:
        data: Raw CBOR bytes to decode.

    Returns:
        Dict mapping integer keys to byte string values.

    Raises:
        ValueError: If the CBOR structure does not match the peace-payload schema.
    """
    m = cbor2.loads(data)
    if not isinstance(m, dict):
        raise ValueError(f"Expected CBOR map, got {type(m).__name__}")
    if 0 not in m:
        raise ValueError("Missing required field 0 (locator)")
    for k, v in m.items():
        if not isinstance(k, int):
            raise ValueError(f"All keys must be int, got {type(k).__name__}")
        if not isinstance(v, bytes):
            raise ValueError(f"All values must be bytes, got {type(v).__name__} for key {k}")
    return m

File: app/gui/fe/src/services/crypto/payload.ts
/**
 * Peace Protocol Standardized Encrypted Payload
 *
 * Builds and parses CBOR-encoded payloads following the peace-payload CDDL schema:
 *   { 0 => bstr, ? 1 => bstr, ? 2 => bstr, * int => bstr }
 *
 * Uses canonical CBOR encoding (RFC 8949 §4.2) via cborg to produce
 * byte-identical output to Python's cbor2 with canonical=True.
 *
 * Copyright (C) 2025 Logical Mechanism LLC
 * SPDX-License-Identifier: GPL-3.0-only
 */

import * as cborg from 'cborg';
import { hexToBytes, bytesToHex } from './bls12381';

/**
 * Structured payload fields for building a peace-payload.
 */
export interface PayloadFields {
  /** Field 0 (required): content address — IPFS CID, Arweave TX ID, URL, or inline data */
  locator: Uint8Array;
  /** Field 1 (optional): access/decryption key for off-chain content */
  secret?: Uint8Array;
  /** Field 2 (optional): integrity hash of the underlying content */
  digest?: Uint8Array;
  /** Fields 3+ (optional): application-specific extension fields */
  extra?: Map<number, Uint8Array>;
}

/**
 * Build a canonical CBOR-encoded peace-payload map.
 *
 * @param fields - Structured payload fields
 * @returns Canonical CBOR bytes (deterministic encoding)
 * @throws Error if extra contains reserved keys (0, 1, 2)
 */
export function buildPayload(fields: PayloadFields): Uint8Array {
  const map = new Map<number, Uint8Array>();
  map.set(0, fields.locator);
  if (fields.secret !== undefined) {
    map.set(1, fields.secret);
  }
  if (fields.digest !== undefined) {
    map.set(2, fields.digest);
  }
  if (fields.extra) {
    for (const [k, v] of fields.extra) {
      if (k === 0 || k === 1 || k === 2) {
        throw new Error(`Extra key ${k} conflicts with reserved keys (0, 1, 2)`);
      }
      map.set(k, v);
    }
  }
  return cborg.encode(map);
}

/**
 * Parse a CBOR-encoded peace-payload map.
 *
 * Validates that the decoded value is a map with integer keys and
 * Uint8Array values, and that field 0 (locator) is present.
 *
 * @param data - Raw CBOR bytes
 * @returns Map of integer keys to Uint8Array values
 * @throws Error if structure doesn't match peace-payload schema
 */
export function parsePayload(data: Uint8Array): Map<number, Uint8Array> {
  const decoded = cborg.decode(data, { useMaps: true });
  if (!(decoded instanceof Map)) {
    throw new Error(`Expected CBOR map, got ${typeof decoded}`);
  }
  if (!decoded.has(0)) {
    throw new Error('Missing required field 0 (locator)');
  }
  for (const [k, v] of decoded) {
    if (typeof k !== 'number') {
      throw new Error(`All keys must be int, got ${typeof k}`);
    }
    if (!(v instanceof Uint8Array)) {
      throw new Error(`All values must be bytes, got ${typeof v} for key ${k}`);
    }
  }
  return decoded as Map<number, Uint8Array>;
}

/**
 * Build a peace-payload from hex-encoded field values.
 *
 * @param fields - Object with hex-encoded field values
 * @returns Hex-encoded canonical CBOR
 */
export function buildPayloadHex(fields: {
  locator: string;
  secret?: string;
  digest?: string;
}): string {
  return bytesToHex(
    buildPayload({
      locator: hexToBytes(fields.locator),
      secret: fields.secret ? hexToBytes(fields.secret) : undefined,
      digest: fields.digest ? hexToBytes(fields.digest) : undefined,
    })
  );
}

/**
 * Parse a hex-encoded CBOR peace-payload into hex-encoded fields.
 *
 * @param hex - Hex-encoded CBOR bytes
 * @returns Object with hex-encoded field values
 */
export function parsePayloadHex(hex: string): {
  locator: string;
  secret?: string;
  digest?: string;
  extra?: Record<number, string>;
} {
  const m = parsePayload(hexToBytes(hex));
  const result: {
    locator: string;
    secret?: string;
    digest?: string;
    extra?: Record<number, string>;
  } = {
    locator: bytesToHex(m.get(0)!),
  };
  if (m.has(1)) result.secret = bytesToHex(m.get(1)!);
  if (m.has(2)) result.digest = bytesToHex(m.get(2)!);
  for (const [k, v] of m) {
    if (k >= 3) {
      if (!result.extra) result.extra = {};
      result.extra[k] = bytesToHex(v);
    }
  }
  return result;
}
